lib/core/waitForCoreChainLockSyncFactory.js

```diff
@@ -0,0 +1,70 @@
+const { ChainLock } = require('@dashevo/dashcore-lib');
+const ZMQClient = require('@dashevo/dashd-zmq');
+const waitForEvent = require('../util/waitForEvent');
+const ensureBlock = require('../util/ensureBlock');
+
+/**
+ * Check that core is synced (factory)
+ *
+ * @param {ZMQClient} coreZMQClient
+ * @param {RpcClient} coreRpcClient
+ * @param {LatestCoreChainLock} latestCoreChainLock
+ * @param {Logger} logger
+ *
+ * @returns {waitForCoreSync}
+ */
+function waitForCoreChainLockSyncFactory(
+  coreZMQClient,
+  coreRpcClient,
+  latestCoreChainLock,
+  logger,
+) {
+  /**
+   * Check that chainLock is synced
+   *
+   * @typedef waitForCoreChainLockSync
+   *
+   * @returns {Promise<LatestCoreChainLock>}
+   */
+  async function waitForCoreChainLockSync() {
+    await coreZMQClient.connect();
+    // listens to ChainLock zmq zmqpubrawchainlocksig event and updates the latestCoreChainLock.
+    // If we lost connection with Core (heartbeats with retry policy) we need to exit with fatal
+    // In case of regtest fallback we need to listen for new blocks using hashblock zmq event
+    coreZMQClient.subscribe(ZMQClient.TOPICS.rawchainlock);
+    logger.info('Subscribe to rawchainlock ZMQ room');
+
+    coreZMQClient.on(ZMQClient.TOPICS.rawchainlock, async (rawChainLockMessage) => {
+      const socketChainLock = new ChainLock(rawChainLockMessage);
+      logger.info(`Received new chainlock ${socketChainLock.toString()}`);
+      await ensureBlock(coreZMQClient, coreRpcClient, socketChainLock.blockHash);
+      latestCoreChainLock.update(socketChainLock);
+    });
+
+    // Because a Chainlock may happen before its block, we also subscribe to rawblock
+    coreZMQClient.subscribe(ZMQClient.TOPICS.hashblock);
+    logger.info('Subscribe to hashblock ZMQ room');
+
+    // We need to retrieve latest chainlock from our fully synced DashCore instance
+    try {
+      const rpcBestChainLockResponse = await coreRpcClient.getBestChainLock();
+      const chainlock = new ChainLock(rpcBestChainLockResponse.result);
+
+      await ensureBlock(coreZMQClient, coreRpcClient, chainlock.blockHash);
+      latestCoreChainLock.update(chainlock);
+    } catch (e) {
+      // Unable to find any chainlock
+      if (e.code === -32603) {
+        logger.info('Waiting for a new chainlock');
+        // We need to wait for new chainlock from ZMQ socket
+        await waitForEvent(coreZMQClient, ZMQClient.TOPICS.rawchainlock);
```

Comment:
It waits when an event happens but it doesn't wait until the listener above will finish. I feel we need a slightly different logic here.

Reply:
When you end up here, it means that rpcBestChainLockResponse has errored into a -32603 error code, so you never `ensureBlock`, when you got to `ensureBlock`, you just throw error (because error won't be -32603), so both condition are not met in the same flow.